strings/check_anagram.nim

```diff
@@ -0,0 +1,179 @@
+                                        ## mapping from Char to a counter slot
+
+const
+  UpperAlpha = {'A'..'Z'}
+  LowerAlpha = {'a'..'z'}
+
+func toLowerUnchecked(c: char): char {.inline.} =
+  ## Lowers the case of an ASCII uppercase letter ('A'..'Z').
+  ## No checks performed, the caller must ensure the input is a valid character.
+  ##
+  ## See also:
+  ## * `strutils.toLowerAscii <https://nim-lang.org/docs/strutils.html#toLowerAscii%2Cchar>`_
+  assert c in UpperAlpha
+  # The difference between numerical values for chars of uppercase and
+  # lowercase alphabets in the ASCII table is 32. Uppercase letters start from
+  # 0b100_0001, so setting the sixth bit to 1 gives letter's lowercase pair.
+  char(uint8(c) or 0b0010_0000'u8)
+
+template normalizeChar(c: char) =
+  ## Checks if the character is a letter and lowers its case.
+  ##
+  ## Raises a `ValueError` on other characters.
+  if c in LowerAlpha: c
+  elif c in UpperAlpha: toLowerUnchecked(c)
+  else: raise newException(ValueError, "Character '" & c & "' is not a letter!")
+
+func isAnagram*(a, b: openArray[char]): bool {.raises: ValueError.} =
+  ## Checks if two words are anagrams of one another.
+  ##
+  ## Raises a `ValueError` on any non-letter character.
+  if a.len != b.len: return false
+  var seenDifferent = false
+  var mapA, mapB: Map
+  for chIdx in 0..<a.len:
+    let (chA, chB) = (normalizeChar(a[chIdx]), normalizeChar(b[chIdx]))
+    # Identical characters xor to 0, must meet at least one 1 (difference)
+    seenDifferent = bool(ord(seenDifferent) or (ord(chA) xor ord(chB)))
+    mapA[chA].inc()
+    mapB[chB].inc()
+  # words are not identical and letter counters match
+  seenDifferent and (mapA == mapB)
+
+func isAnagram(a, bNormalized: openArray[char]; bMap: Map): bool {.raises: ValueError.} =
+  ## Checks if two words are anagrams of one another.
+  ## Leverages a prepared (lowercased) word `bNormalized` and its pre-calculated
+  ## map `bMap` for one-to-many comparisons.
+  ##
+  ## Raises a `ValueError` on any non-letter character.
+  if a.len != bNormalized.len: return false
+  var seenDifferent = false
+  var aMap: Map
+  for chIdx in 0..<a.len:
+    let (chA, chB) = (normalizeChar(a[chIdx]), bNormalized[chIdx])
+    # Identical characters xor to 0, must meet at least one 1 (difference)
+    seenDifferent = bool(ord(seenDifferent) or (ord(chA) xor ord(chB)))
+    aMap[chA].inc()
+  # words are not identical and letter counters match
+  seenDifferent and (aMap == bMap)
+
+func filterAnagrams*(word: string; candidates: openArray[string]): seq[string] {.raises: ValueError.} =
+  ## Checks if any of the candidates is an anagram of a `word` and returns them.
+  ##
+  ## Raises a `ValueError` on any non-letter character.
+  var wordMap: Map
+  var wordNormalized = newString(word.len)
+  for chIdx, c in word.pairs():
+    let c = normalizeChar(c)
+    wordNormalized[chIdx] = c
+    wordMap[c].inc()
```

Comment:
Ok I might relax a little bit the rule on single-character name for string algorithms only.
`character` is a pain to write. It shortens to `char` but it is a keyword (a pre-defined type). `charac` is very long and this variable is used a lot in string algorithms.

Reply:
Well, there's just 5 entities in this block. Young kids, sleepy people and even I can keep 5 things in their head, at least for 4 lines. :D

I'd say, conventional `c` for char, `n` for number and `i` are acceptable when there's no competing variables in current scope.